presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -806,6 +802,26 @@ public void finishInsert(ConnectorSession session, ConnectorInsertTableHandle in
             for (CompletableFuture<?> fileRenameFuture : fileRenameFutures) {
                 MoreFutures.getFutureValue(fileRenameFuture, PrestoException.class);
             }
+
+            // clean remaining partition directories
+            Optional<Path> writePathRoot = locationService.writePath(handle.getLocationHandle(), Optional.empty());
+            if (writePathRoot.isPresent() && !writePathRoot.get().equals(locationService.targetPathRoot(handle.getLocationHandle()))) {
+                for (PartitionUpdate partitionUpdate : partitionUpdates) {
+                    if (writePathRoot.isPresent() && !partitionUpdate.getTargetPath().equals(partitionUpdate.getWritePath())) {
```

Comment:
`writePathRoot` and `targetPathRoot` are either the same or different. One can not be the parent of another (Yes, this should be better documented.) As a result, checking whether `partitionUpdate.getTargetPath()` and `partitionUpdate.getWritePath()` are the same brings up the question of when they can be the same, which can confuse future readers of this code. I would do `Verify.verify` on this condition instead.